File: BE/src/main/java/com/gdgoc/babi_order/clienterror/entity/ClientError.java
package com.gdgoc.babi_order.clienterror.entity;

import com.gdgoc.babi_order.clienterror.ClientErrorSource;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.Instant;

@Entity
@Table(
        name = "client_errors",
        indexes = {
                @Index(name = "idx_client_errors_created_at", columnList = "created_at"),
                @Index(name = "idx_client_errors_tracking_request_id", columnList = "tracking_request_id"),
                @Index(name = "idx_client_errors_related_request_id", columnList = "related_request_id")
        }
)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ClientError {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** POST /api/client-errors 요청 자체의 requestId */
    @Column(name = "tracking_request_id", length = 64)
    private String trackingRequestId;

    /** 연관 Backend API requestId */
    @Column(name = "related_request_id", length = 64)
    private String relatedRequestId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ClientErrorSource source;

    @Column(name = "error_name", nullable = false, length = 200)
    private String errorName;

    @Column(nullable = false, length = 2000)
    private String message;

    @Column(length = 8000)
    private String stack;

    @Column(name = "component_stack", length = 8000)
    private String componentStack;

    @Column(nullable = false, length = 500)
    private String route;

    @Column(name = "user_agent", length = 500)
    private String userAgent;

    @Column(length = 100)
    private String browser;

    @Column(length = 100)
    private String platform;

    @Column(name = "reported_at", nullable = false)
    private Instant reportedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    public ClientError(
            String trackingRequestId,
            String relatedRequestId,
            ClientErrorSource source,
            String errorName,
            String message,
            String stack,
            String componentStack,
            String route,
            String userAgent,
            String browser,
            String platform,
            Instant reportedAt
    ) {
        this.trackingRequestId = trackingRequestId;
        this.relatedRequestId = relatedRequestId;
        this.source = source;
        this.errorName = errorName;
        this.message = message;
        this.stack = stack;
        this.componentStack = componentStack;
        this.route = route;
        this.userAgent = userAgent;
        this.browser = browser;
        this.platform = platform;
        this.reportedAt = reportedAt;
    }

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = Instant.now();
        }
    }
}

File: FE/src/types/clientError.ts
/** Frontend reporter 내부 source (소문자) */
export type FrontendErrorSource =
  | "window"
  | "unhandledrejection"
  | "react"
  | "api";

/** Backend ClientErrorSource enum JSON 값 */
export type BackendClientErrorSource =
  | "WINDOW"
  | "UNHANDLED_REJECTION"
  | "REACT"
  | "API";

/** POST /api/client-errors 요청 body */
export interface ClientErrorPayload {
  timestamp: string;
  source: BackendClientErrorSource;
  errorName: string;
  message: string;
  stack?: string;
  componentStack?: string;
  route: string;
  /** 연관 Backend API 요청의 X-Request-Id (tracking 요청 자체 ID와 별개) */
  relatedRequestId?: string;
  userAgent?: string;
  browser?: string;
  platform?: string;
}

/** reportFrontendError 입력 */
export interface FrontendErrorInput {
  source: FrontendErrorSource;
  error?: unknown;
  message?: string;
  errorName?: string;
  stack?: string;
  componentStack?: string;
  route?: string;
  relatedRequestId?: string;
}

/** ApiError 등에 부착 가능한 연관 requestId */
export interface RelatedRequestIdCarrier {
  relatedRequestId?: string;
}

export const CLIENT_ERROR_API_PATH = "/api/client-errors";

export const REQUEST_ID_HEADER = "X-Request-Id";

export const FRONTEND_TO_BACKEND_SOURCE: Record<
  FrontendErrorSource,
  BackendClientErrorSource
> = {
  window: "WINDOW",
  unhandledrejection: "UNHANDLED_REJECTION",
  react: "REACT",
  api: "API",
};
